### PR TITLE
KAFKA-10604: The StreamsConfig.STATE_DIR_CONFIG's default value does not reflect the JVM parameter or OS-specific settings

### PR DESCRIPTION
**Pull request by dongjinleekr**
Make the default state store directory location to follow OS-specific temporary directory settings or `java.io.tmpdir` JVM parameter, with `Utils#getTempDir`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by mjsax — APPROVED** [review]
Sorry for dropping this on the floor. I guess it looks good. However, the PR needs a rebase.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: Thanks for this @dongjinleekr !

It's kind of a bummer that we actually create this file while defining the configurations, since it results in leaking temp directories, even if users set this config to a different value.

Do you think we can do this in a lazy fashion?

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: Good point John. Maybe we should just use `System.getProperty("java.io.tmpdir")` here?

We have already code in place in `StateDirectory` constructor to create the `./kafka-streams` sub-directory. Might be good enough without any additional changes?

Comment (dongjinleekr) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: @vvcephei @mjsax Totally agree. Since I am now finalizing KAFKA-8403 now, I will fix this with a integration test after that. 🙏

**Comment by dongjinleekr** [comment]
@vvcephei @mjsax Fixed, with rebasing onto the latest trunk and validataion. (please see `StreamsConfigTest.shouldStateDirStartsWithJavaIOTmpDir`.)

**Comment by mjsax** [comment]
@dongjinleekr -- the PR shows merge conflicts. Can you rebase once more. Sorry about that.

**Comment by vvcephei** [comment]
Hey @dongjinleekr ,

Sorry for the force-push, but I had to rebase this and resolve a conflict before merging. Note that the conflict was from 462c89e0b436abd56864bea8bbcaf1ab70b7f66e, which re-organized the boolean conditions in the StateDirectory constructor, specifically where we warn if the state dir is a temp dir.

After resolving the conflict, I noticed there's no test for that warning, so I added one to be sure it works. It also looked like the temp dir check could actually be a bit simpler, so I just tweaked it rather than leaving a new comment for you to address.

I hope this is all ok. I'll let the tests run and merge in the morning, unless you have any objections.

Thanks!
-John

**Comment by vvcephei** [comment]
Ah, a checkstyle violation. How embarrassing. Just fixed it.

**Comment by vvcephei** [comment]
Only one flaky test failure:
```
 Build / JDK 11 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationSSLTest.testOneWayReplicationWithAutoOffsetSync()
```

Proceeding to merge...

**Comment by vvcephei** [comment]
Merged!

Many thanks, @dongjinleekr !

**Pull request merged**
